### PR TITLE
Introduced gaps per monitor - resolved #74

### LINKED ISSUE #74: `gaps.outer.top` vs the notch
Long, _long_ time Yabai user here trying out AeroSpace. After about one day I'm pretty sure I'm a convert. Brilliant work @nikitabobko!

That said, I have a question regarding `gaps.outer.top`. It seems that on the display with a "notch", the padding is in relation to the bottom of the notch. Whereas displays without a notch, the padding is in relation to the actual top of the display. This makes the gap between the top of a window and my bar (SketchyBar) drastically larger/smaller on each display. Is there any way to account for this (eg. display specific setting)?

Maybe a screenshot will help illustrate better than my description. The left side is my MacBook Pro and the right is my external.

CleanShot 2023-12-29 at 18 24 53@2x

### PR DESCRIPTION
**Pull request by wojciech-kulik**
Resolved issue #74

Introduced a new syntax for gaps (based on https://github.com/nikitabobko/AeroSpace/pull/98#issuecomment-1890935012):

```toml
# Gaps between windows (inner-*) and between monitor edges (outer-*).
# Possible values:
# - Constant:         outer-left = 8
# - Per monitor:      outer-left = [{ monitor.main = 16 }, { monitor."some-pattern" = 32 }, 24]
#   In this example, 24 is a default value when there is no match.

[gaps]
inner-horizontal = 10
inner-vertical = [{ monitor."main" = 1 }, { monitor."secondary" = 2 }, 5]
outer-left = 12
outer-bottom = 13
outer-top = [{ monitor."built-in" = 3 }, { monitor."secondary" = 4 }, 6]
outer-right = [{ monitor.2 = 7 }]
```

Notes:
I tried to implement it in an extensible way. I introduced `DynamicConfigValue<T>` which can be used to provide either a constant value or an array of rules.

Currently, this type defines two possible values: `constant` and `perMonitor(rules, defaultValue)`. However, it could be extended in the future to support more conditions.

PS. I also fixed some Xcode warnings.

### CODE REVIEW

**Comment by wojciech-kulik** [comment]
This PR introduces breaking changes. 
What should be done in this case? 
Do you assume that breaking changes can appear in the beta phase?

I guess the best approach would be to start releasing new versions regularly with appropriate release notes.

**Review by nikitabobko — CHANGES_REQUESTED** [review]

File: src/config/parseConfig.swift [hunk 1]
```diff
@@ -237,6 +237,61 @@ func parseString(_ raw: TOMLValueConvertible, _ backtrace: TomlBacktrace) -> Par
...
+        return .failure(.semantic(backtrace, "Missing \"monitor\" table"))
+    }
+
+    return table.lazy.compactMap { key, value in
+        let backtrace = backtrace + .key("monitor") + .key(key)
+        let monitorDescriptionResult = parseMonitorDescription(key, backtrace)
+
+        guard let monitorDescription = monitorDescriptionResult.getOrNil() else {
+            return .failure(monitorDescriptionResult.errorOrNil ?? .semantic(backtrace, "Invalid monitor rule"))
+        }
```
[the remark this answers is not in the document]
Comment (wojciech-kulik) [reply src/config/parseConfig.swift:285]: What should be returned then by the function?

File: src/config/parseConfig.swift [hunk 2]
```diff
@@ -237,6 +237,61 @@ func parseString(_ raw: TOMLValueConvertible, _ backtrace: TomlBacktrace) -> Par
...
+
+        let rules: [ParsedToml<PerMonitorValue<T>>] = array.compactMap {
+            parsePerMonitorValue($0, backtrace)
+        }
+
+        guard !rules.isEmpty else {
+            return .failure(.semantic(backtrace, "Rules array must not be empty"))
+        }
+
+        if let error = rules.compactMap(\.errorOrNil).first {
```
[the remark this answers is not in the document]
Comment (wojciech-kulik) [reply src/config/parseConfig.swift:258]: What should I return if there are errors when my output is non-optional and non-`Result`?

**Comment by nikitabobko** [comment]
Merged with the following patch:
```diff
diff --git a/LocalPackage/Sources/Common/util/SequenceEx.swift b/LocalPackage/Sources/Common/util/SequenceEx.swift
index 4ce8b24..8c231b1 100644
--- a/LocalPackage/Sources/Common/util/SequenceEx.swift
+++ b/LocalPackage/Sources/Common/util/SequenceEx.swift
@@ -15,6 +15,11 @@ public extension Sequence {
         return result
     }
 
+    var first: Element? {
+        var iter = makeIterator()
+        return iter.next()
+    }
+
     func mapAllOrFailure<T, E>(_ transform: (Self.Element) -> Result<T, E>) -> Result<[T], E> {
         var result: [T] = []
         for element in self {
diff --git a/docs/config-examples/default-config.toml b/docs/config-examples/default-config.toml
index e0422dd..5b4579f 100644
--- a/docs/config-examples/default-config.toml
+++ b/docs/config-examples/default-config.toml
@@ -34,26 +34,25 @@ default-root-container-layout = 'tiles'
 #               tall monitor (anything higher than wide) gets vertical orientation
 default-root-container-orientation = 'auto'
 
+# Gaps between windows (inner-*) and between monitor edges (outer-*).
+# Possible values:
+# - Constant:     gaps.outer.top = 8
+# - Per monitor:  gaps.outer.top = [{ monitor.main = 16 }, { monitor."some-pattern" = 32 }, 24]
+#                 In this example, 24 is a default value when there is no match.
+#                 Monitor pattern is the same as for 'workspace-to-monitor-force-assignment'.
+#                 See: https://nikitabobko.github.io/AeroSpace/guide.html#assign-workspaces-to-monitors
+gaps.inner.horizontal = 0
+gaps.inner.vertical =   0
+gaps.outer.left =       0
+gaps.outer.bottom =     0
+gaps.outer.top =        0
+gaps.outer.right =      0
+
 # Visual indent makes it easier to understand that containers of the same orientation are nested.
 # If you have 'enable-normalization-opposite-orientation-for-nested-containers' enabled then
 # there is no way to observe the indent
 indent-for-nested-containers-with-the-same-orientation = 30
 
-# Gaps between windows (inner-*) and between monitor edges (outer-*).
-# Possible values:
-# - Constant:         outer-left = 8
-# - Per monitor:      outer-left = [{ monitor.main = 16 }, { monitor."some-pattern" = 32 }, 24]
-#   In this example, 24 is a default value when there is no match.
-#   Monitor pattern is the same as for 'workspace-to-monitor-force-assignment'.
-#   See: https://nikitabobko.github.io/AeroSpace/guide.html#assign-workspaces-to-monitors
-[gaps]
-inner.horizontal = 0
-inner.vertical =   0
-outer.left =       0
-outer.bottom =     0
-outer.top =        0
-outer.right =      0
-
 # 'main' binding mode declaration
 # See: https://nikitabobko.github.io/AeroSpace/guide.html#binding-modes
 # 'main' binding mode must be always presented
diff --git a/src/config/parseConfig.swift b/src/config/parseConfig.swift
index 235b0fc..62fdac7 100644
--- a/src/config/parseConfig.swift
+++ b/src/config/parseConfig.swift
@@ -27,7 +27,7 @@ private func getConfigFileUrl() -> URL? {
         ?? FileManager.default.homeDirectoryForCurrentUser.appending(path: ".config/")
     let candidates = [
         FileManager.default.homeDirectoryForCurrentUser.appending(path: dotFileName),
-        xdgConfigHome.appending(path: "aerospace").appending(path: fileName)
+        xdgConfigHome.appending(path: "aerospace").appending(path: fileName),
     ]
     let existingCandidates: [URL] = candidates.filter { (candidate: URL) in FileManager.default.fileExists(atPath: candidate.path) }
     let count = existingCandidates.count
@@ -88,8 +88,8 @@ typealias ParsedToml<T> = Result<T, TomlParseError>
 
 extension ParserProtocol {
     func transformRawConfig(_ raw: S,
-        _ value: TOMLValueConvertible,
-        _ backtrace: TomlBacktrace,
+                            _ value: TOMLValueConvertible,
+                            _ backtrace: TomlBacktrace,
                             _ errors: inout [TomlParseError]) -> S {
         raw.copy(keyPath, parse(value, backtrace, &errors).getOrNil(appendErrorTo: &errors))
     }
@@ -244,78 +244,78 @@ func parseSimpleType<T>(_ raw: TOMLValueConvertible) -> T? {
 func parseDynamicValue<T>(
     _ raw: TOMLValueConvertible,
     _ valueType: T.Type,
-    _ initial: T,
+    _ fallback: T,
     _ backtrace: TomlBacktrace,
     _ errors: inout [TomlParseError]
 ) -> DynamicConfigValue<T> {
     if let simpleValue = parseSimpleType(raw) as T? {
         return .constant(simpleValue)
     } else if let array = raw.array {
-        guard !array.isEmpty else {
-            errors.append(.semantic(backtrace, "Array must not be empty"))
-            return .constant(initial)
+        if array.isEmpty {
+            errors.append(.semantic(backtrace, "The array must not be empty"))
+            return .constant(fallback)
         }
 
         guard let defaultValue = array.last.flatMap({ parseSimpleType($0) as T? }) else {
-            errors.append(.semantic(backtrace, "The last item must be a default value (type: \(T.self))"))
-            return .constant(initial)
+            errors.append(.semantic(backtrace, "The last item in the array must be of type \(T.self)"))
+            return .constant(fallback)
         }
 
-        let rules: [PerMonitorValue<T>] = parsePerMonitorValues(TOMLArray(array.dropLast()), backtrace, &errors)
-
-        guard !rules.isEmpty else {
-            errors.append(.semantic(backtrace, "Array must containt at least one monitor pattern"))
-            return .constant(initial)
+        if array.dropLast().isEmpty {
+            errors.append(.semantic(backtrace, "The array must contain at least one monitor pattern"))
+            return .constant(fallback)
         }
 
+        let rules: [PerMonitorValue<T>] = parsePerMonitorValues(TOMLArray(array.dropLast()), backtrace, &errors)
+
         return .perMonitor(rules, default: defaultValue)
     } else {
         errors.append(.semantic(backtrace, "Unsupported type: \(raw.type), expected: \(valueType) or array"))
-        return .constant(initial)
+        return .constant(fallback)
     }
 }
 
 func parsePerMonitorValues<T>(_ array: TOMLArray, _ backtrace: TomlBacktrace, _ errors: inout [TomlParseError]) -> [PerMonitorValue<T>] {
-    array.enumerated().compactMap { index, raw -> PerMonitorValue<T>? in
+    array.enumerated().compactMap { (index: Int, raw: TOMLValueConvertible) -> PerMonitorValue<T>? in
         var backtrace = backtrace + .index(index)
 
-        guard let tableItem = raw.table else {
-            errors.append(expectedActualTypeError(expected: .table, actual: raw.type, backtrace))
+        guard let (key, value) = raw.unwrapTableWithSingleKey(expectedKey: "monitor", &backtrace)
+            .flatMap({ $0.value.unwrapTableWithSingleKey(expectedKey: nil, &backtrace) })
+            .getOrNil(appendErrorTo: &errors) else {
             return nil
         }
 
-        guard tableItem.keys.count == 1 else {
-            errors.append(.semantic(backtrace, "Table must contain exactly one key with monitor table"))
-            return nil
-        }
-
-        guard let table = tableItem["monitor"]?.table else {
-            errors.append(.semantic(backtrace, "Missing monitor table"))
-            return nil
-        }
+        let monitorDescriptionResult = parseMonitorDescription(key, backtrace)
 
-        backtrace = backtrace + .key("monitor")
+        guard let monitorDescription = monitorDescriptionResult.getOrNil(appendErrorTo: &errors) else { return nil }
 
-        guard table.keys.count == 1 else {
-            errors.append(.semantic(backtrace, "Monitor table must contain exactly one key with monitor pattern"))
+        guard let value = parseSimpleType(value) as T? else {
+            errors.append(.semantic(backtrace, "Expected type is '\(T.self)'. But actual type is '\(value.type)'"))
             return nil
         }
 
-        return table.lazy.compactMap { key, value in
-            let backtrace = backtrace + .key(key)
-            let monitorDescriptionResult = parseMonitorDescription(key, backtrace)
-
-            guard let monitorDescription = monitorDescriptionResult.getOrNil(appendErrorTo: &errors) else {
-                return nil
-            }
-
-            guard let value = parseSimpleType(value) as T? else {
-                errors.append(.semantic(backtrace, "Expected type is '\(T.self)'. But actual type is '\(value.type)'"))
-                return nil
-            }
+        return (description: monitorDescription, value: value) as PerMonitorValue<T>
+    }
+}
 
-            return (description: monitorDescription, value: value) as PerMonitorValue<T>
-        }.first
+private extension TOMLValueConvertible {
+    func unwrapTableWithSingleKey(expectedKey: String? = nil, _ backtrace: inout TomlBacktrace) -> ParsedToml<(key: String, value: TOMLValueConvertible)> {
+        guard let table = table else {
+            return .failure(expectedActualTypeError(expected: .table, actual: type, backtrace))
+        }
+        let singleKeyError: TomlParseError = .semantic(backtrace,
+            expectedKey != nil
+                ? "The table is expected to have a single key '\(expectedKey!)'"
+                : "The table is expected to have a single key"
+        )
+        guard let (actualKey, value): (String, TOMLValueConvertible) = table.count == 1 ? table.first : nil else {
+            return .failure(singleKeyError)
+        }
+        if expectedKey != nil && expectedKey != actualKey  {
+            return .failure(singleKeyError)
+        }
+        backtrace = backtrace + .key(actualKey)
+        return .success((actualKey, value))
     }
 }
 
@@ -519,7 +519,7 @@ indirect enum TomlBacktrace: CustomStringConvertible {
         }
     }
 
-    static func + (lhs: TomlBacktrace, rhs: TomlBacktrace) -> TomlBacktrace {
+    static func +(lhs: TomlBacktrace, rhs: TomlBacktrace) -> TomlBacktrace {
         if case .root = lhs {
             if case .key(let newRoot) = rhs {
                 return .rootKey(newRoot)
diff --git a/src/config/parseGaps.swift b/src/config/parseGaps.swift
index 890513a..0555665 100644
--- a/src/config/parseGaps.swift
+++ b/src/config/parseGaps.swift
@@ -3,19 +3,19 @@ import TOMLKit
 
 private let gapsParser: [String: any ParserProtocol<RawGaps>] = [
     "inner": Parser(\.inner, parseInner),
-    "outer": Parser(\.outer, parseOuter)
+    "outer": Parser(\.outer, parseOuter),
 ]
 
 private let innerParser: [String: any ParserProtocol<RawInner>] = [
     "vertical": Parser(\.vertical) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
-    "horizontal": Parser(\.horizontal) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) }
+    "horizontal": Parser(\.horizontal) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
 ]
 
 private let outerParser: [String: any ParserProtocol<RawOuter>] = [
     "left": Parser(\.left) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
     "bottom": Parser(\.bottom) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
     "top": Parser(\.top) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
-    "right": Parser(\.right) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) }
+    "right": Parser(\.right) { value, backtrace, errors in parseDynamicValue(value, Int.self, 0, backtrace, &errors) },
 ]
 
 func parseGaps(_ raw: TOMLValueConvertible, _ backtrace: TomlBacktrace, _ errors: inout [TomlParseError]) -> Gaps {
diff --git a/test/config/ConfigTest.swift b/test/config/ConfigTest.swift
index 7593d5e..5919f4d 100644
--- a/test/config/ConfigTest.swift
+++ b/test/config/ConfigTest.swift
@@ -276,7 +276,7 @@ final class ConfigTest: XCTestCase {
     }
 
     func testParseGaps() {
-        let (config, errors) = parseConfig(
+        let (config, errors1) = parseConfig(
             """
             [gaps]
             inner.horizontal = 10
@@ -287,7 +287,7 @@ final class ConfigTest: XCTestCase {
             outer.right = [{ monitor.2 = 7 }, 8]
             """
         )
-        XCTAssertEqual(errors.descriptions, [])
+        XCTAssertEqual(errors1.descriptions, [])
         XCTAssertEqual(
             config.gaps,
             Gaps(
@@ -303,5 +303,18 @@ final class ConfigTest: XCTestCase {
                 )
             )
         )
+
+        let (_, errors2) = parseConfig(
+            """
+            [gaps]
+            inner.horizontal = [true]
+            inner.vertical = [{ foo.main = 1 }, { monitor = { foo = 2, bar = 3 } }, 1]
+            """
+        )
+        XCTAssertEqual(errors2.descriptions, [
+            "gaps.inner.horizontal: The last item in the array must be of type Int",
+            "gaps.inner.vertical[0]: The table is expected to have a single key \'monitor\'",
+            "gaps.inner.vertical[1].monitor: The table is expected to have a single key",
+        ])
     }
 }
diff --git a/test/testExtensions.swift b/test/testExtensions.swift
index 4231693..acb031b 100644
--- a/test/testExtensions.swift
+++ b/test/testExtensions.swift
@@ -4,7 +4,7 @@ import Foundation
 import TOMLKit
 
 extension WindowDetectedCallback: Equatable {
-    public static func == (lhs: WindowDetectedCallback, rhs: WindowDetectedCallback) -> Bool {
+    public static func ==(lhs: WindowDetectedCallback, rhs: WindowDetectedCallback) -> Bool {
         check(
             lhs.matcher.appNameRegexSubstring == nil &&
                 lhs.matcher.windowTitleRegexSubstring == nil &&
@@ -21,19 +21,19 @@ extension [TomlParseError] {
 }
 
 extension Mode: Equatable {
-    public static func == (lhs: Mode, rhs: Mode) -> Bool {
+    public static func ==(lhs: Mode, rhs: Mode) -> Bool {
         lhs.name == rhs.name && lhs.bindings == rhs.bindings
     }
 }
 
 extension HotkeyBinding: Equatable {
-    public static func == (lhs: HotkeyBinding, rhs: HotkeyBinding) -> Bool {
+    public static func ==(lhs: HotkeyBinding, rhs: HotkeyBinding) -> Bool {
         lhs.modifiers == rhs.modifiers && lhs.key == rhs.key && lhs.commands.map(\.describe) == rhs.commands.map(\.describe)
     }
 }
 
 extension MonitorDescription: Equatable {
-    public static func == (lhs: MonitorDescription, rhs: MonitorDescription) -> Bool {
+    public static func ==(lhs: MonitorDescription, rhs: MonitorDescription) -> Bool {
         switch (lhs, rhs) {
         case (.sequenceNumber(let a), .sequenceNumber(let b)):
             return a == b
@@ -50,7 +50,7 @@ extension MonitorDescription: Equatable {
 }
 
 extension DynamicConfigValue: Equatable {
-    public static func == (lhs: Self, rhs: Self) -> Bool {
+    public static func ==(lhs: Self, rhs: Self) -> Bool {
         switch (lhs, rhs) {
         case (.constant(let lhsConstant), .constant(let rhsConstant)):
             return lhsConstant == rhsConstant
@@ -65,20 +65,20 @@ extension DynamicConfigValue: Equatable {
 }
 
 extension Gaps: Equatable {
-    public static func == (lhs: Self, rhs: Self) -> Bool {
+    public static func ==(lhs: Self, rhs: Self) -> Bool {
         lhs.outer == rhs.outer && lhs.inner == rhs.inner
     }
 }
 
 extension Gaps.Outer: Equatable {
-    public static func == (lhs: Self, rhs: Self) -> Bool {
+    public static func ==(lhs: Self, rhs: Self) -> Bool {
         lhs.left == rhs.left && lhs.right == rhs.right &&
             lhs.bottom == rhs.bottom && lhs.top == rhs.top
     }
 }
 
 extension Gaps.Inner: Equatable {
-    public static func == (lhs: Self, rhs: Self) -> Bool {
+    public static func ==(lhs: Self, rhs: Self) -> Bool {
         lhs.horizontal == rhs.horizontal && lhs.vertical == rhs.vertical
     }
 }
```

**Comment by nikitabobko** [comment]
https://github.com/nikitabobko/AeroSpace/commit/377c35bbbe1febf3d89738e524d91cda11ca4a6f

**Comment by wojciech-kulik** [comment]
🍻 

PS. Out of curiosity, why don't you use pull requests to merge/push changes?

**Pull request closed without merging**